notes/plutus-core/Bytestrings.md

```diff
@@ -0,0 +1,116 @@
+(19th November 2020)
+
+In a [Slack thread](https://input-output-rnd.slack.com/archives/C21UF2WVC/p1599120765047200)
+some time ago, Alex asked about extending support for bytestrings in Plutus Core.
+This document describes the issue and suggests some possible extensions.
+
+
+### Current bytestring support in PLC
+
+Plutus Core currently provides the following operations for its built-in
+bytestring type:
+
+   * `concatenate`
+   * `equalsByteString`
+   * `lessThanByteString`
+   * `greaterThanByteString`
+   * `takeByteString`
+   * `dropByteString`
+   * `sha2_256`
+   * `sha3_256`
+   * `verifySignature`
+
+The `concatenate` function might be better called `appendByteString`.  The PLC
+specification also mentions an `emptyByteString` builtin, but we don't support
+that.  However we do provide concrete syntax for bytestrings:
+`(con bytestring #91ef22)` for example; `(con bytestring #)` represents the empty bytestring.
+
+The problem is that there is no easy way to _construct_ a bytestring in Haskell
+and convert it into a PLC bytestring, or to programatically construct a new
+bytestring inside a Plutus Core program.  The `Data` type supports bytestrings
+and can be used to pass bytestrings into validators, but this isn't very
+general-purpose.
+
+
+### Characters and strings in Plutus Core
+
+Plutus Core also provides built-in character and string types.  These both have
+concrete syntactical representations: `(con char 'q')` and `(con string
+"something")`.  Support for these is very limited: we only have `charToString`
+and `append` (append one string to another); also the PlutusTx compiler can
+compile literal Haskell strings and characters into PLC values of the
+appropriate types (but see below for some complications involving Strings).
+
+However there is a subtlety: Plutus Core actually has _two_ representations of
+Haskell Strings.  The PlutusTx plugin compiles normal Haskell strings into
+Scott-encoded lists of PLC's `char` type ("Scott strings" for short), but there
+is also a special type called `Language.PlutusTx.Builtins.String` which is
+compiled into PLCs' `string` type.  Many standard string operations can be
+compiled into PLC code acting on Scott strings, but support for builtin strings
+in PlutusTx is limited.
+
+### Bytestrings in Haskell
+
+Plutus Core bytestrings are implemented using Haskell's (strict) `ByteString`
+type from `Data.ByteString`.  Internally these are implemented as arrays of
+`Word8` values.  Haskell doesn't provide any concrete syntax for the
+`ByteString` type, but there are functions `pack :: [Word8] -> ByteString` and
+`unpack :: ByteString -> [Word8]` which convert back and forth between
+`ByteString`s and lists of `Word8` values.
+
+One way (and possbily the most efficient way) to improve support for bytestrings
+in Plutus Core would be to add the `Word8` type to the default PLC universe, but
+this would add quite a lot of complexity that we probably don't want.  [However,
```

Comment:
Why?

Reply:
Why not `Word8`?  I'm not opposed to it in principle, but then we might want comparisons, arithmetic operations, etc.  That could increase the number of builtins quite significantly, but maybe that's OK.

Reply:
Or maybe we can sneak type classes into `TypeScheme` and use `add` for all numeric types. Not sure if I'm joking or not.